server/src/main/java/org/elasticsearch/repositories/RepositoriesService.java

```diff
@@ -126,30 +128,52 @@ public void registerRepository(final PutRepositoryRequest request, final ActionL
         final RepositoryMetadata newRepositoryMetadata = new RepositoryMetadata(request.name(), request.type(), request.settings());
         validate(request.name());
 
-        final ActionListener<AcknowledgedResponse> registrationListener;
+        // Trying to create the new repository on master to make sure it works
+        try {
+            closeRepository(createRepository(newRepositoryMetadata, typesRegistry));
+        } catch (Exception e) {
+            listener.onFailure(e);
+            return;
+        }
+
+        final StepListener<AcknowledgedResponse> acknowledgementStep = new StepListener<>();
+        final StepListener<Void> publicationStep = new StepListener<>();
+
         if (request.verify()) {
-            registrationListener = ActionListener.delegateFailure(listener, (delegatedListener, clusterStateUpdateResponse) -> {
+
+            // When publication has completed (and all acks received or timed out) then verify the repository.
+            // (if acks timed out then acknowledgementStep completes before the master processes this cluster state, hence why we have
+            // to wait for the publication to be complete too)
+            final StepListener<List<DiscoveryNode>> verifyStep = new StepListener<>();
+            publicationStep.whenComplete(ignored -> acknowledgementStep.whenComplete(clusterStateUpdateResponse -> {
                 if (clusterStateUpdateResponse.isAcknowledged()) {
                     // The response was acknowledged - all nodes should know about the new repository, let's verify them
-                    verifyRepository(request.name(), delegatedListener.map(discoveryNodes -> clusterStateUpdateResponse));
+                    verifyRepository(request.name(), verifyStep);
                 } else {
-                    delegatedListener.onResponse(clusterStateUpdateResponse);
+                    verifyStep.onResponse(null);
                 }
-            });
-        } else {
-            registrationListener = listener;
-        }
+            }, listener::onFailure), listener::onFailure);
 
-        // Trying to create the new repository on master to make sure it works
-        try {
-            closeRepository(createRepository(newRepositoryMetadata, typesRegistry));
-        } catch (Exception e) {
-            registrationListener.onFailure(e);
-            return;
+            // When verification has completed, get the repository data for the first time
+            final StepListener<RepositoryData> getRepositoryDataStep = new StepListener<>();
+            verifyStep.whenComplete(ignored -> threadPool.generic().execute(
+                    ActionRunnable.wrap(getRepositoryDataStep, l -> repository(request.name()).getRepositoryData(l))), listener::onFailure);
+
+            // When the repository metadata is ready, update the repository UUID stored in the cluster state, if available
```

Comment:
There is one slightly strange spot here for the case of read-only repositories.
You could have technically have a repo at some UUID and mount it to a cluster read-only. Then some outside force could wipe the repo completely (as in delete all the files) and write new `RepositoryData` with a different UUID, in which case the uuid in the metadata won't ever get fixed. It's kind of a weird edge-case but we it's the reason why we don't track the repo generation in the cluster state when it comes to read-only repositories as well.

I guess in this case we explicitly want this UUID visible in read-only repos also so maybe we need some check on the UUID when loading `RepositoryData` on read-only repos?

Reply:
Ugh I see, ok, should be nbd since `getRepositoryData` is already async.

Reply:
See b680eb3.